docs/building-a-form/about-the-json-schema-standard-and-react-jsonschema-form.md

```diff
@@ -0,0 +1,296 @@
+
+The schemaform code uses rjsf to render form fields, but it builds a scaffolding on top of it to support multi-page forms and our common form patterns. The second part is creating a form configuration spec that allows devs to specify the structure of one of our multi-page forms.
+
+##### Customizing fields, widgets, and events from rjsf
+
+rjsf passes all field and widget components to `SchemaField` (and most other components) as a `registry` property. To override fields and widgets in the registry, pass components of the same name into the library's main `Form` component. The schemaform uses custom versions of these components:
+
+- `ObjectField`
+- `ArrayField`
+- `FieldTemplate`
+- `TextWidget`
+- `SelectWidget`
+- `EmailWidget`
+- `CheckboxWidget`
+- `RadioWidget`
+- `TextareaWidget`
+
+The schemaform uses these custom fields and widgets:
+
+- `YesNoWidget`
+- `AddressField`
+- `DateWidget`
+- `SSNWidget`
+- `PhoneWidget`
+
+Writing custom widgets is similar to writing React components: A value is passed in, and an `onChange` hook is provided for changing data. Other properties like the schemas and field ID are also provided.
+
+Custom fields receive all properties listed previously for field components in rjsf.
+
+In addition to customizing fields and widgets, the schemaform code hooks into a number of events provided by `Form` to support our form patterns, found in the `FormPage` component. These events are:
+
+- `validate`: This event is called when validation occurs. We call our custom validation, which reads uiSchema for custom validation hooks that have been included for form fields outside of what JSON Schema provides.
+- `transformErrors`: This event is provided when the schemaform receives the list of JSON Schema validation errors and can return a transformed list. It replaces the messages with a set of default messages, as well as any messages provided for specific fields in uiSchema. It also moves the errors for required fields from the object level to the field level. Because JSON Schema specifies required fields with a `required` array on an object field schema, any errors about missing data are associated with that object and moved so they're associated with the missing field and rendered with that field on the form.
+- `onError`: This event is called if a user tries to submit a form with a validation error. The schemaform sets a `submitted` flag in `formContext`, which is an object passed to all fields and components in the rjsf form. The `FieldTemplate` component uses `formContext` to display all error messages to the user.
+- `onSubmit`: This event is called when a user submits a form with no validation errors. When this happens, the schemaform code looks for the next page in the multi-page form and navigates to it.
+- `onChange`: This event is called when a user changes data in the form. The schemaform fires a Redux action and updates the store with the new data. The reducer code does several recalculations:
```

Comment:
It's not clear to me when this occurs vis-a-vis the section above titled "About the onChange field component".